File: frontend/src/lib/richtext.ts
/**
 * 書式付きの文字(richtext 型)の扱い。値は HTML で持つ。
 * 描くときは、ここで許した要素と属性だけを残す(サーバも同じ規則で保存前に洗う前提。04 §1)。
 */
const ALLOWED: Record<string, string[]> = {
  p: [],
  br: [],
  strong: [],
  b: [],
  em: [],
  i: [],
  u: [],
  s: [],
  ul: [],
  ol: [],
  li: [],
  h1: [],
  h2: [],
  h3: [],
  blockquote: [],
  code: [],
  pre: [],
  a: ['href'],
}

/** @ で言及したレコード。エディタ(Tiptap の Mention)が書く形と同じ */
export interface Mention {
  object: string
  id: string
  label: string
}

const MENTION_ID = /^([a-z][a-z0-9_]*):([0-9a-f-]{36})$/

export function parseMentionId(raw: string | null | undefined): { object: string; id: string } | null {
  const m = raw ? MENTION_ID.exec(raw) : null
  return m ? { object: m[1], id: m[2] } : null
}

/** 内容(HTML)から言及を拾う。同じレコードは 1 回。サーバも同じ規則で mentions 列を作る */
export function extractMentions(html: string | null | undefined): Mention[] {
  if (!html) return []
  // 表示と同じく DOM で解釈する(正規表現だと引用符の違いで取りこぼす)
  const doc = new DOMParser().parseFromString(`<body>${html}</body>`, 'text/html')
  const out: Mention[] = []
  const seen = new Set<string>()
  for (const el of Array.from(doc.querySelectorAll('[data-type="mention"]'))) {
    const id = el.getAttribute('data-id')
    const ref = parseMentionId(id)
    if (!ref || seen.has(id!)) continue
    seen.add(id!)
    out.push({ ...ref, label: el.getAttribute('data-label') ?? el.textContent?.replace(/^@/, '') ?? '' })
  }
  return out
}

function clean(node: Node, doc: Document): Node | null {
  if (node.nodeType === Node.TEXT_NODE) return doc.createTextNode(node.textContent ?? '')
  if (node.nodeType !== Node.ELEMENT_NODE) return null
  const el = node as Element
  const tag = el.tagName.toLowerCase()
  // 言及は <span data-type="mention" data-id="テーブル名:ID" data-label="名前">@名前</span> だけを通す
  if (tag === 'span' && el.getAttribute('data-type') === 'mention' && parseMentionId(el.getAttribute('data-id'))) {
    const out = doc.createElement('span')
    out.setAttribute('data-type', 'mention')
    out.setAttribute('data-id', el.getAttribute('data-id')!)
    out.setAttribute('data-label', el.getAttribute('data-label') ?? '')
    out.setAttribute('role', 'link')
    out.setAttribute('tabindex', '0')
    out.textContent = el.textContent ?? ''
    return out
  }
  const allowed = ALLOWED[tag]
  // 許さない要素は、中身だけを残す(<div> や <span> の入れ子を平らにする)
  const out = allowed ? doc.createElement(tag) : doc.createDocumentFragment()
  if (allowed && out instanceof Element) {
    for (const name of allowed) {
      const value = el.getAttribute(name)
      if (name === 'href' && value && /^(https?:|mailto:|tel:)/i.test(value.trim())) {
        out.setAttribute('href', value)
        out.setAttribute('target', '_blank')
        out.setAttribute('rel', 'noreferrer')
      }
    }
  }
  for (const child of Array.from(el.childNodes)) {
    const c = clean(child, doc)
    if (c) out.append(c)
  }
  return out
}

export function sanitizeHtml(html: string): string {
  const doc = new DOMParser().parseFromString(`<body>${html}</body>`, 'text/html')
  const root = doc.createElement('div')
  for (const child of Array.from(doc.body.childNodes)) {
    const c = clean(child, doc)
    if (c) root.append(c)
  }
  return root.innerHTML
}

/** 書式を落とした素の文字(一覧の要約、検索、CSV に使う) */
export function plainText(html: string): string {
  return html
    .replace(/<\/(p|li|h[1-6]|blockquote|pre|div)>/gi, '\n')
    .replace(/<br\s*\/?>/gi, '\n')
    .replace(/<[^>]+>/g, '')
    .replace(/&nbsp;/g, ' ')
    .replace(/&amp;/g, '&')
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')
    .replace(/&quot;/g, '"')
    .replace(/&#39;/g, "'")
    .replace(/\n{2,}/g, '\n')
    .trim()
}

/** 中身が空か(<p></p> だけの HTML も空とみなす) */
export function isEmptyHtml(html: string | null | undefined): boolean {
  return !html || plainText(html) === ''
}

File: backend/app/records/richtext.py
"""書式付きの文字(`richtext`)の洗浄と、`@` の言及の取り出し。

規則は画面の `frontend/src/lib/richtext.ts` と同じ。**保存前に洗い、その結果から言及を取る**(この順。04 §1)。
洗浄そのものは `nh3` に任せる(自前で書かない。`bleach` は非推奨なので使わない)。
"""

import json
import re
from html.parser import HTMLParser
from typing import Any

import nh3

ALLOWED_TAGS = {
    "p", "br", "strong", "b", "em", "i", "u", "s", "ul", "ol", "li",
    "h1", "h2", "h3", "blockquote", "code", "pre", "a", "span",
}  # fmt: skip
ALLOWED_ATTRIBUTES = {
    "a": {"href"},
    # 言及だけが span を使う(mention でない span は下の _strip_plain_spans が剥がす)
    "span": {"data-type", "data-id", "data-label", "role", "tabindex"},
}
URL_SCHEMES = {"http", "https", "mailto", "tel"}
MENTION_ID = re.compile(r"^([a-z][a-z0-9_]*):([0-9a-f-]{36})$")

VOID_TAGS = {"br"}


def parse_mention_id(raw: str | None) -> tuple[str, str] | None:
    matched = MENTION_ID.match(raw) if raw else None
    return (matched.group(1), matched.group(2)) if matched else None


class _StripPlainSpans(HTMLParser):
    """言及でない `<span>` を、中身だけ残して剥がす(モックの `clean` と同じ)。"""

    def __init__(self) -> None:
        super().__init__(convert_charrefs=True)
        self.out: list[str] = []
        self.stripped: list[bool] = []

    def handle_starttag(self, tag: str, attrs: list[tuple[str, str | None]]) -> None:
        values = dict(attrs)
        if tag == "span" and not (values.get("data-type") == "mention" and parse_mention_id(values.get("data-id"))):
            if tag not in VOID_TAGS:
                self.stripped.append(True)
            return
        if tag not in VOID_TAGS:
            self.stripped.append(False)
        rendered = "".join(f' {name}="{_escape(value or "")}"' for name, value in attrs)
        self.out.append(f"<{tag}{rendered}>")

    def handle_startendtag(self, tag: str, attrs: list[tuple[str, str | None]]) -> None:
        rendered = "".join(f' {name}="{_escape(value or "")}"' for name, value in attrs)
        self.out.append(f"<{tag}{rendered}>")

    def handle_endtag(self, tag: str) -> None:
        if tag in VOID_TAGS:
            return
        if self.stripped and self.stripped.pop():
            return
        self.out.append(f"</{tag}>")

    def handle_data(self, data: str) -> None:
        self.out.append(_escape(data, quote=False))


def _escape(value: str, quote: bool = True) -> str:
    out = value.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;")
    return out.replace('"', "&quot;") if quote else out


def sanitize_html(html: str) -> str:
    """許した要素と属性だけを残す。許さない要素は**中身だけ**残す。"""
    cleaned = nh3.clean(
        html,
        tags=ALLOWED_TAGS,
        attributes=ALLOWED_ATTRIBUTES,
        url_schemes=URL_SCHEMES,
        link_rel="noreferrer",
        set_tag_attribute_values={"a": {"target": "_blank"}, "span": {"role": "link", "tabindex": "0"}},
    )
    parser = _StripPlainSpans()
    parser.feed(cleaned)
    parser.close()
    return "".join(parser.out)


class _Mentions(HTMLParser):
    def __init__(self) -> None:
        super().__init__(convert_charrefs=True)
        self.found: list[dict[str, str]] = []
        self.seen: set[str] = set()

    def handle_starttag(self, tag: str, attrs: list[tuple[str, str | None]]) -> None:
        if tag != "span":
            return
        values = dict(attrs)
        if values.get("data-type") != "mention":
            return
        raw = values.get("data-id")
        parsed = parse_mention_id(raw)
        if not parsed or raw in self.seen:
            return
        self.seen.add(str(raw))
        self.found.append({"object": parsed[0], "id": parsed[1], "label": values.get("data-label") or ""})


def extract_mentions(html: str | None) -> list[dict[str, str]]:
    """内容から言及を拾う。同じレコードは 1 回。**洗浄したあとの HTML に対して行う**(04 §1)。"""
    if not html:
        return []
    parser = _Mentions()
    parser.feed(html)
    parser.close()
    return parser.found


def mentions_value(html: str | None) -> str | None:
    """レコードに載せる形(`テーブル名:ID` の JSON)。正は `body` で、これは捨てて作り直せる導出値。"""
    ids = [f"{m['object']}:{m['id']}" for m in extract_mentions(html)]
    return json.dumps(ids, ensure_ascii=False, separators=(",", ":")) if ids else None


def body_field(obj: dict[str, Any]) -> str | None:
    timeline = obj.get("timeline")
    return timeline["body"] if timeline else None
